server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -305,18 +313,48 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
         }
     }
 
-    public void copyFrom(Directory from, String src, String dest, IOContext context, boolean useCommonSuffix) throws IOException {
+    public String getMetadataFileAndAcquireLock(long primaryTerm, long generation, String resourceID) throws IOException {
+        Collection<String> metadataFiles = remoteMetadataDirectory.listFilesByPrefix(
+            MetadataFilenameUtils.METADATA_PREFIX);
+
+        Optional<String> metadataFile = metadataFiles.stream().filter(
+            file -> (
+                MetadataFilenameUtils.getPrimaryTerm(
+                    file.split(MetadataFilenameUtils.SEPARATOR)) == primaryTerm
+                && MetadataFilenameUtils.getGeneration(
+                    file.split(MetadataFilenameUtils.SEPARATOR)) == generation)).findAny();
+
+        if (metadataFile.isEmpty()) {
+            String errorString = "Metadata file is not present for given primary term "
+                + primaryTerm + " and generation " + generation;
+            throw new FileNotFoundException(errorString);
+        }
+        mdLockManager.acquireLock(metadataFile.get(), resourceID);
+        return metadataFile.get();
+    }
+
+    public void copyFrom(Directory from, String src, String dest, IOContext context, boolean useCommonSuffix,
+                         String checksum) throws IOException {
         String remoteFilename;
         if (useCommonSuffix) {
             remoteFilename = dest + SEGMENT_NAME_UUID_SEPARATOR + this.commonFilenameSuffix;
         } else {
             remoteFilename = getNewRemoteSegmentFilename(dest);
         }
         remoteDataDirectory.copyFrom(from, src, remoteFilename, context);
-        String checksum = getChecksumOfLocalFile(from, src);
+        if (checksum == null) {
+            checksum = getChecksumOfLocalFile(from, src);
+        }
         UploadedSegmentMetadata segmentMetadata = new UploadedSegmentMetadata(src, remoteFilename, checksum);
         segmentsUploadedToRemoteStore.put(src, segmentMetadata);
     }
+    public void copyFrom(Directory from, String src, String dest, IOContext context, boolean useCommonSuffix) throws IOException {
+        copyFrom(from, src, dest, context, useCommonSuffix, null);
```

Comment:
instead of passing `null` you can just pass `getChecksumOfLocalFile(from, src)` from this caller. Then the null handling will not be needed in other method.